integrations/clickup/client.py

```diff
@@ -0,0 +1,231 @@
+        yield (await self._send_api_request(f"{self.api_url}/team")).get("teams", [])
+
+    @cache_iterator_result()
+    async def _get_spaces_in_team(
+        self, team_id: str
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all spaces in a workspace."""
+        yield (
+            await self._send_api_request(
+                f"{self.api_url}/team/{team_id}/space",
+                {"is_archived": self.is_archived},
+            )
+        ).get("spaces", [])
+
+    @cache_iterator_result()
+    async def _get_folders_in_space(
+        self, team_id: str
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all folders in a space."""
+        async for spaces in self._get_spaces_in_team(team_id):
+            for space in spaces:
+                yield (
+                    await self._send_api_request(
+                        f"{self.api_url}/space/{space.get('id')}/folder",
+                        {"is_archived": self.is_archived},
+                    )
+                ).get("folders")
+
+    async def get_folder_projects(self) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all projects with a folder parent."""
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for folders in self._get_folders_in_space(team.get("id")):
+                    for folder in folders:
+                        projects = folder.get("lists")
+                        yield [{**project, TEAM_OBJECT: team} for project in projects]
+
+    async def get_folderless_projects(
+        self,
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all projects without a folder parent."""
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for spaces in self._get_spaces_in_team(team.get("id")):
+                    for space in spaces:
+                        response = await self._send_api_request(
+                            f"{self.api_url}/space/{space.get('id')}/list",
+                            {"is_archived": self.is_archived},
+                        )
+                        projects = response.get("lists")
+                        yield [{**project, TEAM_OBJECT: team} for project in projects]
+
+    async def get_single_project(self, list_id: str) -> Optional[Dict[str, Any]]:
+        """Get a single project by list_id."""
+        url = f"{self.api_url}/list/{list_id}"
+        response = await self._send_api_request(url)
+        space_id = response.get("space").get("id")
+
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for spaces in self._get_spaces_in_team(team.get("id")):
+                    for space in spaces:
+                        logger.info(space.get("id"))
+                        if space.get("id") == space_id:
+                            return {**response, TEAM_OBJECT: team}
```

Comment:
I have an idea would like to get your opinion on it
what if we will have the `space` blueprint as well as mapping, we will not have to do all of that effort and amounts of requests to get the project.

https://clickup.com/api/clickupreference/operation/GetSpace/

we will have the following relations

issue -> project -> space -> team

inside the project blueprint we will have the `team` as a mirror from the space relation, that way, we will not have to concatenate the `team_object` in each project entity and therefor reduce a lot of redundant requests.

cc @oiadebayo @PeyGis

Reply:
This has been done